types/underscore/underscore-tests.ts

```diff
@@ -165,372 +165,423 @@ declare const level2UnionList: _.List<_.List<string | number>>;
+    { title: "Other", author: "Not Shakespeare", year: 2012 }
+],
+    { author: "Shakespeare", year: 1611 });
 
 // determining whether every value is truthy
-_.every([true, 1, null, 'yes']);
-
-_.any([null, 0, 'yes', false]);
+_.every([true, 1, null, 'yes']); // $ExpectType boolean
 
 // determining whether any number in a list is divisible by three
-_.some([1, 2, 3, 4], l => l % 3 === 0);
+_.some(numberArray, l => l % 3 === 0); // $ExpectType boolean
 
 // determining whether any value in a dictionary is uppercase
-_.some({ a: 'a', b: 'B', c: 'C', d: 'd' }, l => l === l.toUpperCase());
+_.some({ a: 'a', b: 'B', c: 'C', d: 'd' }, l => l === l.toUpperCase()); // $ExpectType boolean
 
 // checking whether an item is in an array
-_.contains([1, 2, 3], 3);
+_.contains(numberArray, 3); // $ExpectType boolean
 
 // checking whether an item is in the portion of an array that starts with index 1
-_.contains([1, 2, 3], 3, 1);
+_.contains(numberArray, 3, 1); // $ExpectType boolean
 
 // truncating a set of strings to 5 characters or less
-_.invoke(['zebra', 'giraffe', 'lion'], 'substring', 0, 5);
+_.invoke(['zebra', 'giraffe', 'lion'], 'substring', 0, 5); // $ExpectType any[]
 
 // retrieving a property value from all items in a collection
-var stooges = [{ name: 'moe', age: 40 }, { name: 'larry', age: 50 }, { name: 'curly', age: 60 }];
-_.pluck(stooges, 'name');
+_.pluck(stooges, 'name'); // $ExpectType string[]
 
 // retrieving the minimum number in a dictionary
-_.min({ a: 1, b: 2 });
+_.min(numberDictionary); // $ExpectType number
 
 // retrieving the item with the maximum number in a property
-_.max(stooges, (stooge) => stooge.age);
+_.max(stooges, (stooge) => stooge.age); // $ExpectType number | { name: string; age: number; }
 
 // sorting by a calculated value
-_.sortBy([1, 2, 3, 4, 5, 6], (num) => Math.sin(num));
+_.sortBy(numberArray, num => Math.sin(num)); // $ExpectType number[]
 
 // grouping numbers by their non-fractional parts
-_([1.3, 2.1, 2.4]).groupBy((e) => Math.floor(e));
+_([1.3, 2.1, 2.4]).groupBy((e) => Math.floor(e)); // $ExpectType Dictionary<number[]>
 
 // grouping numbers by the value of a specified property
-_.groupBy(['one', 'two', 'three'], 'length');
+_.groupBy(['one', 'two', 'three'], 'length'); // $ExpectType Dictionary<string[]>
 
 // indexing items in a dictionary by age
-_.indexBy(stooges, 'age');
+_.indexBy(stooges, 'age'); // $ExpectType Dictionary<{ name: string; age: number; }>
 
 // counting numbers by their evenness
-_.countBy([1, 2, 3, 4, 5], (num) => (num % 2 == 0) ? 'even' : 'odd');
+_.countBy(numberArray, num => (num % 2 === 0) ? 'even' : 'odd'); // $ExpectType Dictionary<number>
 
 // shuffling numbers
-_.shuffle([1, 2, 3, 4, 5, 6]);
+_.shuffle(numberArray); // $ExpectType number[]
 
 // converting an array-like structure to an array
-(function (a, b, c, d) { return _.toArray(arguments).slice(1); })(1, 2, 3, 4);
+(function (a, b, c, d) {
+    const args = _.toArray(arguments); // $ExpectType any[]
```

Comment:
I'm debating possibly exchanging this test for something simpler like `_.toArray(explicitNumberDictionary); // $ExpectType number[]` since the resulting type in this example isn't very interesting and I feel like rest parameters make the use of `arguments` in this fashion obsolete at this point; thoughts?

Reply:
It's a good point for existing code in particular that ideally this should still be prevented from breaking since folks shouldn't be forced to move away from this in existing code; I'll keep it but perhaps add a new test as well.

I wasn't thinking about this from a TS-specific standpoint but rather from a JS standpoint in general, but I suppose when types aren't involved and in particular overloads or [labeled tuple elements](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#labeled-tuple-elements) aren't available for documentation one would probably want to write a function with defined parameter names in JS for documentation purposes but not painstakingly build an array that is a combination of all of the parameters provided for documentation purposes, and folks may want to migrate their JS to TS without needing to rewrite patterns like those to use a single rest parameter.

i.e. in TS prior to 4.0 I might write
```TS
function foo(format: string, mode: number, ...formatArgs: string[]);
function foo(...args: [string, number, ...string[]]) {
    doSomethingWithAllTheArgs(args);
}
```

but in JS I'd probably write
```JS
function foo(format, mode, ...formatArgs) {
    doSomethingWithAllTheArgs(_.toArray(arguments));
}
```